app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,270 @@
+        if (++tap_dance->counter >= cfg->behavior_count) {
+            press_tap_dance_behavior(tap_dance, event.timestamp);
+        }
+    } else {
+        tap_dance = store_tap_dance(event.position, cfg);
+        if (tap_dance == NULL) {
+            LOG_ERR("unable to store tap dance, did you press more than %d tap_dance?",
+                    ZMK_BHV_TAP_DANCE_MAX_HELD);
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+    }
+    tap_dance->is_pressed = true;
+    // No other key was pressed. Start the timer.
+    tap_dance->release_at = event.timestamp + tap_dance->config->tapping_term_ms;
+    // adjust timer in case this behavior was queued by a hold-tap
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    LOG_DBG("ms_left equal to: %d", ms_left);
+    if (ms_left > 0) {
+        k_delayed_work_submit(&tap_dance->release_timer, K_MSEC(ms_left));
+    }
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int on_tap_dance_binding_released(struct zmk_behavior_binding *binding,
+                                         struct zmk_behavior_binding_event event) {
+    LOG_DBG("On Binding Released");
+    struct active_tap_dance *tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        LOG_ERR("ACTIVE TAP DANCE CLEARED TOO EARLY");
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = false;
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    LOG_DBG("ms_left equal to: %d", ms_left);
+    if (ms_left <= 0) {
+        release_tap_dance_behavior(tap_dance, event.timestamp);
+    }
+
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+void behavior_tap_dance_timer_handler(struct k_work *item) {
+    LOG_DBG("Timer Handler Called");
+    struct active_tap_dance *tap_dance = CONTAINER_OF(item, struct active_tap_dance, release_timer);
+    if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
+        return;
+    }
+    if (tap_dance->timer_cancelled) {
+        tap_dance->timer_cancelled = false;
+    } else {
+        if (!tap_dance->is_pressed) {
+            press_tap_dance_behavior(tap_dance, tap_dance->release_at);
+            release_tap_dance_behavior(tap_dance, tap_dance->release_at);
+            return;
+        } else {
+            press_tap_dance_behavior(tap_dance, tap_dance->release_at);
+        }
+    }
+}
+
+static const struct behavior_driver_api behavior_tap_dance_driver_api = {
+    .binding_pressed = on_tap_dance_binding_pressed,
+    .binding_released = on_tap_dance_binding_released,
+};
+
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_tap_dance, tap_dance_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_tap_dance, zmk_position_state_changed);
+
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh) {
+    LOG_DBG("Position state changed");
+    struct zmk_position_state_changed *ev = as_zmk_position_state_changed(eh);
+    if (ev != NULL) {
```

Comment:
if you swap this around and return if `ev == NULL` you don't need as much indentations.